zpl2.py

```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
             stop=self._field_data_stop(),
         )
         self._write_command(command)
+
+    @staticmethod
+    def encode_grf_ascii(pil_image):
+        """Converts a PIL image to a .GRF file encoded in ascii for use
+        with zebra printers.
+        @:return array of strings, each string is a pixel row, containing
+        4 pixel per ascii character
+        """
+
+        def round_up(value, multiple):
+            result = value
+            rest = value % multiple
+            if rest:
+                result += multiple - rest
+            return result
+
+        def get_byte_val(bits, word_len):
+            if len(bits) == 0:
+                return 0
+            else:
+                byte_value = bits[0] << word_len - 1
+                return byte_value + get_byte_val(bits[1:], word_len - 1)
+
+        # make sure we have a monochrome image, and load it.
+        data = pil_image.convert("1").load()
+        sizex, sizey = pil_image.size
+
+        # increase row size to the nearest multiple of 8 pixels
+        # The length of the first row of pixels is set in bytes and must be
+        # a multiple of 8 bytes.
+        # The zebra tools also round the x size of images to the nearest of
+        # 8 pixels.
+        row_size = round_up(sizex, 8)
+        # construct an array y - x
+        # the first row of the array, is the first line of pixels in the
+        # image
+        # Fill all values with white
+        pixels = [[1 for x in range(0, row_size)] for y in range(0, sizey)]
+        for xpos in range(0, sizex):
+            for ypos in range(0, sizey):
+                input_value = data[xpos, ypos]
+                # PIL: pixel value 0 = white , pixel value 255 = black
+                # GRF: 1=white, 0=black
+                bit_value = input_value >> 7 ^ 1
+                pixels[ypos][xpos] = bit_value
+
+        ret = []
+        for pixel_row in pixels:
+            pixel_string = ''
+            for xpos in range(0, len(pixel_row), 4):
+                bits = pixel_row[xpos:xpos + 4]
+                byte_value = get_byte_val(bits, 4)
+                pixel_string += '%X' % byte_value
+            ret.append(pixel_string)
+
+        return ret
+
+    def graphic_image(self, right, down, pil_image):
+        """ Print the image_data from the file handle """
+
+        img_size_y = pil_image.size[1]
+        ascii_data = self.encode_grf_ascii(pil_image)
+
+        # The bytes per row, is the number of chars in the first row, divided
+        # by 8
+        # Should be equal to the X size of the image, divided by 8 and
+        # rounded up
+        bytes_per_row = len(ascii_data[0])/2
```

Comment:
Hmmmm it looks like we're doing this dynamically, but have the `8` hardcoded in the `encode_grf_ascii` method. Am I misunderstanding the intents of the two numbers?